Add auth method for Planetary computer

uclaros opened the pull request:
## Description
This PR adds a new `PlanetaryComputer` auth method for use with the Microsoft Planetary Computer STAC catalog.
The planetary computer requires using SAS sign tokens to access most assets' data, something that was implemented on the old STAC plugin but was missing from the core STAC implementation.

There are two modes for the new auth method:
One is for the Open version of planetary computer, performing the SAS signing for assets:
<img width="610" height="656" alt="image" src="https://github.com/user-attachments/assets/74c30559-07d3-41ef-943b-79cc0ab3d1c5" />

and the second is for Planetary Computer Pro GeoCatalogs, where in addition to the SAS signing of assets, an oauth2 authentication is also required:
<img width="878" height="656" alt="image" src="https://github.com/user-attachments/assets/fa79c530-1d9f-4157-a6e2-61c48a9e7cf3" />


The new auth method can be used with STAC connections, or applied directly to individual gdal or point cloud layers whose data source is on planetary computer.
When a layer is added from a STAC catalog or directly using the data source manager, the `authcfg` will be appended to its URI so QGIS will use that config for the network requests.

Authcfg encoding/decoding was also added to point cloud layers as it was previously ignored even though it was available in the GUI.


<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - you have read the QGIS Developer Guide (https://docs.qgis.org/testing/en/docs/developers_guide/index.html) and your PR complies with its QGIS Coding Standards
-->

uclaros committed:
Add authentication support to remote point cloud sources

uclaros committed:
Add MS Planetary Computer auth method
Supports SAS signing of asset links for the Open version of planetary computer.
Supports oauth2 authorization and SAS signing for planetary computer pro.

uclaros committed:
Propagate STAC connection authcfg to assets
The authcfg is now added to the layer uri and used for asset downloading and preview

nyalldawson commented on src/auth/planetary_computer/core/qgsauthplanetarycomputermethod.cpp:
```diff
@@ -0,0 +1,317 @@
+
+
+const QString QgsAuthPlanetaryComputerMethod::AUTH_METHOD_KEY = QStringLiteral( "PlanetaryComputer" );
+const QString QgsAuthPlanetaryComputerMethod::AUTH_METHOD_DESCRIPTION = QStringLiteral( "MS Planetary Computer" );
+const QString QgsAuthPlanetaryComputerMethod::AUTH_METHOD_DISPLAY_DESCRIPTION = tr( "MS Planetary Computer" );
```
```suggestion
const QString QgsAuthPlanetaryComputerMethod::AUTH_METHOD_DESCRIPTION = QStringLiteral( "Microsoft Planetary Computer" );
const QString QgsAuthPlanetaryComputerMethod::AUTH_METHOD_DISPLAY_DESCRIPTION = tr( "Microsoft Planetary Computer" );
```

nyalldawson commented on src/auth/planetary_computer/core/qgsauthplanetarycomputermethod.cpp:
```diff
@@ -0,0 +1,317 @@
+QgsAuthPlanetaryComputerMethod::QgsAuthPlanetaryComputerMethod()
+{
+  setExpansions( QgsAuthMethod::NetworkRequest | QgsAuthMethod::DataSourceUri );
+  setDataProviders( QStringList() << QStringLiteral( "gdal" ) << QStringLiteral( "copc" ) << QStringLiteral( "stac" ) );
+
```
```suggestion
  setDataProviders( { QStringLiteral( "gdal" ), QStringLiteral( "copc" ), QStringLiteral( "stac" ) } );

```

nyalldawson commented on src/auth/planetary_computer/gui/qgsauthplanetarycomputeredit.cpp:
```diff
@@ -0,0 +1,158 @@
+{
+  setupUi( this );
+
+  connect( cbType, static_cast<void ( QComboBox::* )( int )>( &QComboBox::currentIndexChanged ), this, &QgsAuthPlanetaryComputerEdit::updateServerType ); // also updates GUI
```
```suggestion
  connect( cbType, qOverload< int >( &QComboBox::currentIndexChanged ), this, &QgsAuthPlanetaryComputerEdit::updateServerType ); // also updates GUI
```

nyalldawson commented on src/auth/planetary_computer/gui/qgsauthplanetarycomputeredit.ui:
```diff
@@ -0,0 +1,122 @@
+      </sizepolicy>
+     </property>
+     <property name="text">
+      <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-style:italic;&quot;&gt;The Directory (tenant) and Application (client) IDs can be found in your organization's MS Entra ID main and application page respectively.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
```suggestion
      <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-style:italic;&quot;&gt;The Directory (tenant) and Application (client) IDs can be found in your organization's Microsoft Entra ID main and application page respectively.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```

nyalldawson commented on python/core/auto_generated/pointcloud/qgspointcloudindex.sip.in:
```diff
@@ -170,7 +170,7 @@ concurrent access to the index is memory safe.
 
 
-    void load( const QString &fileName );
+    void load( const QString &fileName, const QString &authcfg = QString() );
```
Add note on when authcfg argument was added

nyalldawson commented on src/core/stac/qgsstacdataitems.cpp:
```diff
@@ -116,6 +118,8 @@ QgsMimeDataUtils::UriList QgsStacItemItem::mimeUris() const
            it->href().startsWith( QLatin1String( "ftp" ), Qt::CaseInsensitive ) )
       {
         uri.uri = QStringLiteral( "/vsicurl/%1" ).arg( it->href() );
+        if ( !authcfg.isEmpty() )
+          uri.uri.append( QStringLiteral( " authcfg='%1'" ).arg( authcfg ) );
```
This isn't a valid URI format for the GDAL provider.

nyalldawson commented on src/gui/stac/qgsstacsourceselect.cpp:
```diff
@@ -623,19 +622,26 @@ void QgsStacSourceSelect::showFootprints( bool enable )
+  const QString authcfg = mStac->authCfg();
+  if ( !authcfg.isEmpty() )
+  {
+    layerUri += QStringLiteral( " authcfg='%1'" ).arg( authcfg );
```
Again, this won't work for gdal provider

uclaros committed:
auth method review fixes

uclaros committed:
Rename params and update docstring in point cloud index

uclaros committed:
Make SasToken::isValid() const

uclaros committed:
remove redundant inner if in QgsAuthPlanetaryComputerMethod

wonder-sk reviewed (commented):
The code is looking good!

uclaros committed:
update auth method help label

uclaros commented:
I've updated the help label for both open and pro versions.
Added a link to microsoft docs (https://learn.microsoft.com/en-us/azure/planetary-computer/get-started-planetary-computer) which should cover the generic administrative steps for any application, and stated that we use a fixed redirect url (http://127.0.0.1:7070) for the oauth token, which is something an admin might need.
I don't think we should explain the required azure internals here as this is targeted mostly to users.

agiudiceandrea commented:
> Interesting... Is it on a STAC connection?

No. It is the dialog window from Settings->Options...->Authentication->Add new authentication configuration

Running QGIS 4.1.0-Master https://github.com/qgis/QGIS/commit/94acfd87483 (from OSGeo4W on the same system) I see the auth method for Planetary computer in the same dialog window, instead.
<img width="573" height="450" alt="image" src="https://github.com/user-attachments/assets/1942cf77-a7e1-43eb-8e46-b6b7cc91cb56" />
